bin/pkg.ml

```diff
@@ -177,14 +183,48 @@ module Lock = struct
         let+ workspace = Memo.run (Workspace.workspace ()) in
         List.filter_map workspace.contexts ~f:(function
           | Workspace.Context.Default
-              { lock; version_preference = version_preference_context; _ } ->
-            Option.map lock ~f:(fun lock_dir_path ->
-                { lock_dir_path
-                ; version_preference =
-                    Version_preference.choose ~from_arg:version_preference_arg
-                      ~from_context:version_preference_context
-                })
+              { lock
+              ; version_preference = version_preference_context
+              ; base = { name = context_name; _ }
+              ; _
+              } ->
+            let lock_dir_path =
+              Option.value lock ~default:Dune_pkg.Lock_dir.default_path
+            in
+            Some
+              { lock_dir_path
+              ; version_preference =
+                  Version_preference.choose ~from_arg:version_preference_arg
+                    ~from_context:version_preference_context
+              ; context_name
+              }
           | Opam _ -> None)
+
+    let contexts_with_dup_lock_dir_paths ts =
+      let context_names_by_lock_dir_path =
+        List.fold_left ts ~init:Path.Source.Map.empty
+          ~f:(fun acc { lock_dir_path; context_name; _ } ->
+            Path.Source.Map.update acc lock_dir_path ~f:(function
+              | None -> Some [ context_name ]
+              | Some context_names -> Some (context_name :: context_names)))
+      in
+      List.find_opt (Path.Source.Map.to_list context_names_by_lock_dir_path)
+        ~f:(fun (_, context_names) -> List.length context_names > 1)
+
+    let check_for_dup_lock_dir_paths ts =
+      match contexts_with_dup_lock_dir_paths ts with
+      | None -> ()
+      | Some (lock_dir_path, context_names) ->
+        User_error.raise
```

Comment:
This error should include the location from the workspace file.

Reply:
This error doesn't necessarily indicate a problem in the workspace file though. If you're just generating one lockdir at a time it's valid (at the moment at least) for all the contexts to use the same lockdir. They will just overwrite each other. Including the location in the error might mislead users to think that they need to change the workspace file.

Alternatively we could make it an error to have multiple contexts with the same lockdir path in general in which case we'd include the location in the error message.

Not sure which solution would be most appropriate.

Reply:
Sure, the error doesn't mean the workspace file is bad, but it's still useful to point at where dune obtained the conflicting information.

Reply:
Fair enough. I just changed it to include the lines from the workspace file spanning the conflicting context definitions. In the process I ran into a problem where `Loc.span` would produce unexpected results when its second argument ended ended before its first argument which I fixed.

Reply:
Not sure if `Loc.span` is the right thing to use since it might include a lot of irrelevant stuff. 

Look for `package_defined_twice` in `dune_project.ml` for an example of how we have two locations in a single error.

Reply:
Using span is still suspicious here